src/module_4/model_fit.py

```diff
@@ -0,0 +1,460 @@
+        figure_filename = f"Evaluation_Curves_{timestamp}.png"
+        figure_path = os.path.join(save_curves_path, figure_filename)
+
+        plt.savefig(figure_path)
+
+    plt.show()
+
+
+def find_precision_threshold(y_val, y_scores, target_precision=0.25):
+    precisions, recalls, thresholds = precision_recall_curve(y_val, y_scores)
+    # Find the closest precision to the target precision
+    idx = np.argmin(np.abs(precisions - target_precision))
+    return thresholds[idx]
+
+
+def train_catboost_model(
+    X_train: pd.DataFrame,
+    y_train: pd.Series,
+    X_val: pd.DataFrame,
+    y_val: pd.Series,
+    params: Dict,
+    X_test: Optional[pd.DataFrame] = None,
+    y_test: Optional[pd.Series] = None,
+) -> CatBoostClassifier:
+    """
+    Trains a CatBoostClassifier model.
+
+    Parameters:
+    - train_data (pd.DataFrame): The training dataset features.
+    - train_labels (pd.Series): The training dataset labels.
+    - val_data (pd.DataFrame): The validation dataset features.
+    - val_labels (pd.Series): The validation dataset labels.
+    - test_data (pd.DataFrame, optional): The test dataset features.
+    - test_labels (pd.Series, optional): The test dataset labels.
+    - params (Dict): Dictionary containing the training parameters,
+    which must include the names of categorical features.
+
+    Returns:
+    - CatBoostClassifier: The trained CatBoost model.
+    """
+
+    # Get the list of categorical feature names and remove it from params
+    # (should not be passed to CatBoostClassifier later on)
+    categorical_features = params.pop("categorical_features", None)
+    categorical_features_indices = [
+        X_train.columns.get_loc(col) for col in categorical_features
+    ]
+
+    # Create CatBoost Pool for train and validation sets
+    train_pool = Pool(X_train, y_train, cat_features=categorical_features_indices)
+    val_pool = Pool(X_val, y_val, cat_features=categorical_features_indices)
+
+    # Optionally create a Pool for the test set
+    test_pool = None  # noqa
+    if X_test is not None and y_test is not None:
+        test_pool = Pool(  # noqa
+            X_test, y_test, cat_features=categorical_features_indices
+        )
+
+    catboost_model = CatBoostClassifier(**params)
+
+    # Train the model
+    catboost_model.fit(train_pool, eval_set=val_pool)
+
+    # Calculate precision threshold
+    threshold = find_precision_threshold(
+        y_val, catboost_model.predict_proba(X_val)[:, 1]
+    )
+
+    return catboost_model, threshold
+
+
+def handler_fit(event: dict) -> dict:
+    """
+    The handler function to load data, train the model, and save it to disk.
+
+    Parameters:
+    - event (dict): Dictionary with input parameters such as path, params, etc.
+
+    Returns:
+    - dict: Dictionary with model information such as model name, path, etc.
+    """
+    try:
+        # Extract parameters from event or set defaults
+        data_path = event.get("data_path")
```

Comment:
ideally you should have these in a config file: https://docs.python.org/3/library/configparser.html#:~:text=A%20configuration%20file%20consists%20of,removed%20from%20keys%20and%20values.